modules/codegen/src/smithy4s/codegen/ToLine.scala

```diff
@@ -18,92 +18,163 @@ package smithy4s.codegen
-      case Primitive.Short      => line"Short"
-      case Primitive.Long       => line"Long"
-      case Primitive.Float      => line"Float"
-      case Primitive.Double     => line"Double"
-      case Primitive.BigDecimal => line"BigDecimal"
-      case Primitive.BigInteger => line"BigInt"
-      case Primitive.Uuid       => Line(Set("java.util.UUID"), "UUID")
-      case Primitive.Document   => Line(Set("smithy4s.Document"), "Document")
-      case Primitive.Nothing    => line"Nothing"
+      case Primitive.Unit       => TypeReference("Unit")
+      case Primitive.ByteArray  => TypeReference("smithy4s", "ByteArray")
+      case Primitive.Bool       => TypeReference("Boolean")
+      case Primitive.String     => TypeReference("String")
+      case Primitive.Timestamp  => TypeReference("smithy4s", "Timestamp")
+      case Primitive.Byte       => TypeReference("Byte")
+      case Primitive.Int        => TypeReference("Int")
+      case Primitive.Short      => TypeReference("Short")
+      case Primitive.Long       => TypeReference("Long")
+      case Primitive.Float      => TypeReference("Float")
+      case Primitive.Double     => TypeReference("Double")
+      case Primitive.BigDecimal => TypeReference("BigDecimal")
+      case Primitive.BigInteger => TypeReference("BigInt")
+      case Primitive.Uuid       => TypeReference("java.util", "UUID")
+      case Primitive.Document   => TypeReference("smithy4s", "Document")
+      case Primitive.Nothing    => TypeReference("Nothing")
     }
 }
 
 // Models
-case class Line(imports: Set[String], line: String) {
-  def tupled = (imports, line)
-  def suffix(suffix: Line) = modify(s => s"$s $suffix")
-  def addImport(imp: String) = copy(imports = imports + imp)
-  def modify(f: String => String) = Line(imports, f(line))
-  def nonEmpty = line.nonEmpty
-  def toLines = Lines(imports, List(line))
-
-  def args(l: LinesWithValue*): Lines = if (l.exists(_.render.lines.nonEmpty)) {
-    val openBlock = if (line.nonEmpty) line + "(" else ""
-    Lines(imports, List(openBlock)) ++ indent(
-      l.toList.foldMap(_.render).mapLines(_ + ",")
-    ) ++ Lines(")")
-  } else Lines(line + "()").addImports(imports)
 
+sealed trait LineSegment { self =>
+  def toLine: Line = Line(Chain.one(self))
+}
+object LineSegment {
+  case class Import(value: String) extends LineSegment
+  case class Hardcoded(value: String) extends LineSegment
+  case class TypeDefinition(name: String) extends LineSegment
+  case class TypeReference(pkg: List[String], name: String)
+      extends LineSegment { self =>
+    def asValue: String = s"${(pkg :+ name).mkString(".")}"
+  }
+  object TypeReference {
+    def apply(pkg: String, name: String): Line =
+      TypeReference(pkg.split("\\.").toList, name).toLine
+    def apply(fqn: String): Line = {
+      val parts = fqn.split("\\.").toList
+      TypeReference(parts.dropRight(1), parts.last).toLine
+    }
+
+  }
+
+  implicit val hardcodedShow = Show.show[Hardcoded](_.value)
```

Comment:
[nitpick] these instances would be better place in the companion objects of their respective types